py-polars/polars/dataframe/frame.py

```diff
@@ -3475,8 +3475,35 @@ def transpose(
         │ 1           ┆ 2           ┆ 3           │
         └─────────────┴─────────────┴─────────────┘
 
-        """
-        df = self._from_pydf(self._df.transpose(include_header, header_name))
+        Use an existing column as the new column names
+
+        >>> df = pl.DataFrame(dict(id=["a", "b", "c"], col1=[1, 3, 2], col2=[3, 4, 6]))
+        >>> df.transpose(column_names="id")
+        shape: (2, 3)
+        ┌─────┬─────┬─────┐
+        │ a   ┆ b   ┆ c   │
+        │ --- ┆ --- ┆ --- │
+        │ i64 ┆ i64 ┆ i64 │
+        ╞═════╪═════╪═════╡
+        │ 1   ┆ 3   ┆ 2   │
+        │ 3   ┆ 4   ┆ 6   │
+        └─────┴─────┴─────┘
+        >>> df.transpose(include_header=True, header_name="new_id", column_names="id")
+        shape: (2, 4)
+        ┌────────┬─────┬─────┬─────┐
+        │ new_id ┆ a   ┆ b   ┆ c   │
+        │ ---    ┆ --- ┆ --- ┆ --- │
+        │ str    ┆ i64 ┆ i64 ┆ i64 │
+        ╞════════╪═════╪═════╪═════╡
+        │ col1   ┆ 1   ┆ 3   ┆ 2   │
+        │ col2   ┆ 3   ┆ 4   ┆ 6   │
+        └────────┴─────┴─────┴─────┘
+        """
+        pydf = self._df
+        if isinstance(column_names, str):
```

Comment:
Can we add this logic to the rust side. That way we can keep that aligned.

Reply:
I did it here because it looks like previously the column naming was only done on the Python side. ...and for some reason someone wanted to be able to pass in generators. If I move it to rust it seems best to move _all_ the column naming there.

Reply:
Yes, I think that is best.  That keeps the functionality aligned and allows other front-ends (R, NodeJS) to have the same functionality.

Reply:
Actually, it looks like _all_ naming of any kind for the transpose is done on the python side. The only rust code that handles some renaming is in py-polars. It would be a much larger task to make this available everywhere. But if you think it's worth it I'll just strip out all the python specific transpose code and move it to rust.

Reply:
I think it is worth it to have all acutal logic on the rust side, but I don't think it should be in this PR per se. So let's merge this first. Thanks!